prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -84,25 +86,198 @@ def map_generator_to_bus_by_sub(generator, bus_groupby):
+    otherwise return NA.
+
+    :param pandas.Series generator: one generating unit from data frame.
+    :param pandas.GroupBy epa_ampd_groupby: data frame of EPA AMPD samples, grouped by
+        plant ID and unit ID.
+    :param pandas.Series crosswalk_translation: mapping of EIA keys to EPA keys.
+    :param int min_unique_x: minimum number of unique x points needed to fit a curve to.
+    :param int min_points: minimum number points needed to fit a curve to.
+    :return: (*tuple*) -- if possible, quadratic coefficients:
+        c0 (constant), c1 (linear), and c2 (quadratic); respectively. Otherwise, return
+        NA values for each
+    """
+
+    def quadratic(x, a, b, c):
+        return a + b * x + c * x ** 2
+
+    quadratic_coefficient_bounds = ([0] * 3, [float("inf")] * 3)
+    return_index = ["h0", "h1", "h2"]
+    default_return = pd.Series([float("nan")] * 3, index=return_index)
+    eia_key = (generator["Plant Code"], generator["Generator ID"])
+    x_col = "GLOAD (MW)"
+    y_col = "HEAT_INPUT (mmBtu)"
+    try:
+        epa_key = crosswalk_translation.loc[eia_key].values[0]
+        samples = epa_ampd_groupby.get_group(epa_key)
+        filtered_samples = samples.query(
+            f"(not `{x_col}`.isnull()) and `{x_col}` != 0 "
+            f"and (not `{y_col}`.isnull()) and `{y_col}` != 0"
+        )
+        if len(filtered_samples[x_col].unique()) < min_unique_x:
+            return default_return
+        if len(filtered_samples) < min_points:
+            return default_return
+        xs = filtered_samples[x_col]
+        ys = filtered_samples[y_col]
+        popt, _ = curve_fit(quadratic, xs, ys, bounds=quadratic_coefficient_bounds)
+        return pd.Series(popt, index=return_index)
+    except KeyError:
+        return default_return
+
+
+def filter_suspicious_heat_rates(generators):
+    """Identify linearized heat rates which violate assumptions of 'reasonable' bounds,
+    and replace them with NaN values.
+
+    :param pandas.DataFrame generators: data frame of generators with fitted heat rates
+        (modified inplace).
+    """
+
+    def get_heat_rate_bounds(generator):
+        """Lookup heat rate bounds for a generator, by either type or type and capacity.
+
+        :param pandas.Series generator: single generator attributes.
+        :return: (*pandas.Series*) -- heat rate bounds ('lower', 'upper').
+        """
+        gen_type = tuple(generator[["Technology", "Prime Mover"]])
+        if gen_type in const.reasonable_heat_rates_by_type:
+            bounds = const.reasonable_heat_rates_by_type[gen_type]
+        elif gen_type in const.reasonable_heat_rates_size_cutoffs:
+            if generator["Pmax"] < const.reasonable_heat_rates_size_cutoffs[gen_type]:
+                key = tuple((*gen_type, "small"))
+            else:
+                key = tuple((*gen_type, "large"))
+            bounds = const.reasonable_heat_rates_by_type_and_size[key]
+        else:
+            bounds = (0, float("inf"))
```

Comment:
Not sure whether it will be cleaner if we do following here:
```python  
gen_type = tuple(generator[["Technology", "Prime Mover"]])
size = ("small",) if generator["Pmax"] < reasonable_heat_rates_size_cutoffs.get(gen_type, 0) else ("large",)
bounds = reasonable_heat_rates_by_type.get(
    gen_type,
    reasonable_heat_rates_by_type_and_size.get(gen_type+size, (0, float("inf")))
)
```

Reply:
Nice, I forgot that you can add tuples together. Line length limits sort of kill the length savings, but we can still make it simpler via your `get` approach. A lightly revised version of what you have, to fit `black`...:
```python
gen_type = tuple(generator[["Technology", "Prime Mover"]])
if gen_type in const.reasonable_heat_rates_by_type:
    bounds = const.reasonable_heat_rates_by_type[gen_type]
size = (
    "small"
    if generator["Pmax"] < reasonable_heat_rates_size_cutoffs.get(gen_type, 0)
    else "large"
)
key = gen_type + (size,)
data = const.reasonable_heat_rates_by_type_and_size.get(key, (0, float("inf")))
return pd.Series(data, index=("lower", "upper"))
```
saves us one line, or
```
gen_type = tuple(generator[["Technology", "Prime Mover"]])
if gen_type in const.reasonable_heat_rates_by_type:
    bounds = const.reasonable_heat_rates_by_type[gen_type]
key = (
    gen_type + ("small",)
    if generator["Pmax"] < reasonable_heat_rates_size_cutoffs.get(gen_type, 0)
    else gen_type + ("large",)
)
data = const.reasonable_heat_rates_by_type_and_size.get(key, (0, float("inf")))
return pd.Series(data, index=("lower", "upper"))
```
saves us two.

Reply:
I went with the second one, which in the end only saves us one line, since `reasonable_heat_rates_size_cutoffs` is a pretty long variable name, but I think is cleaner.